packages/plugin-ext/src/hosted/browser/hosted-plugin.ts

```diff
@@ -112,93 +133,54 @@ export class HostedPluginSupport {
-                    if (!map[host]) {
-                        map[host] = [pluginMetadata];
-                    } else {
-                        map[host].push(pluginMetadata);
-                    }
-                    return map;
-                }, {});
-
-                // create one RPC per host and init.
-                Object.keys(pluginsPerHost).forEach(hostKey => {
-                    const plugins: PluginMetadata[] = pluginsPerHost[hostKey];
-                    let pluginID = hostKey;
-                    if (plugins.length >= 1) {
-                        pluginID = getPluginId(plugins[0].model);
-                    }
-
-                    let rpc = this.rpc.get(hostKey);
-                    if (!rpc) {
-                        rpc = this.createServerRpc(pluginID, hostKey);
-                        setUpPluginApi(rpc, container);
-                        this.rpc.set(hostKey, rpc);
-                    }
-
-                    const hostedExtManager = rpc.getProxy(MAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT);
-                    hostedExtManager.$init({
-                        plugins: plugins,
-                        preferences: getPreferences(this.preferenceProviderProvider, initData.roots),
-                        globalState: initData.globalStates,
-                        workspaceState: initData.workspaceStates,
-                        env: { queryParams: getQueryParameters(), language: navigator.language },
-                        extApi: initData.pluginAPIs
-                    }, confStorage);
-                    this.mainPluginApiProviders.getContributions().forEach(p => p.initialize(rpc!, container));
-                    this.backendExtManagerProxy = hostedExtManager;
-                });
-            }
-
-            // update list with loaded plugins
-            initData.plugins.forEach(value => this.loadedPlugins.add(value.model.id));
-        });
-    }
-
-    private initContributions(pluginsMetadata: PluginMetadata[]): [boolean, boolean] {
-        const result: [boolean, boolean] = [false, false];
-        for (const plugin of pluginsMetadata) {
-            if (plugin.model.entryPoint.frontend) {
-                result[0] = true;
-            } else {
-                result[1] = true;
-            }
-
+        const hostToPlugins = new Map<PluginHost, PluginMetadata[]>();
+        for (const plugin of initData.plugins) {
+            const host = plugin.model.entryPoint.frontend ? 'frontend' : plugin.host;
+            const plugins = hostToPlugins.get(plugin.host) || [];
+            plugins.push(plugin);
+            hostToPlugins.set(host, plugins);
             if (plugin.model.contributes) {
                 this.contributionHandler.handleContributions(plugin.model.contributes);
             }
         }
+        await this.theiaReadyPromise;
+        for (const [host, plugins] of hostToPlugins) {
+            const pluginId = getPluginId(plugins[0].model);
```

Comment:
Since the value of the `hostToPlugin` can be an empty array (see [above](https://github.com/theia-ide/theia/pull/5622/files#diff-aa7bab89f1c76c54d8b56a09d0839d1aR143)), you cannot just get the `model` from the 0<sup>th</sup> index. It is unsafe.

Reply:
The host is always added to the map with at least one plugin.

Reply:
Why is the `|| [];` part? (still from above)

Reply:
when the host is added first time, empty array is created with `|| []'  and on next line it is populated with the plugin

Reply:
Ohh, misread. 😊 Sure makes sense. Thanks!